File: engine/tare/rag/embed.py
"""Les embeddings. Ollama d'abord (gratuit, local), OpenAI en repli.

granite-embedding:278m est le modele de cobol-explorer : 768 dimensions,
fenetre de 512 tokens, servi par Ollama sur la machine. Il ne coute rien et il
n'envoie rien dehors. OpenAI (text-embedding-3-small, 1536 dims) n'est utilise
que si Ollama ne repond pas ou n'a pas le modele — et le rapport de build dit
lequel a servi, parce qu'un index moitie Granite moitie OpenAI ne serait pas
interrogeable : deux espaces vectoriels differents, des distances qui ne veulent
rien dire ensemble.

REGLE 3, ecrite en code ici :

  * un lot qui revient avec moins de vecteurs qu'il n'y avait de textes leve
    EmbedError. On ne complete pas avec des zeros — un vecteur nul a une
    distance cosinus definie, il remonterait dans les resultats, et personne ne
    verrait qu'il ne veut rien dire ;
  * un vecteur de mauvaise dimension leve. Le melange silencieux de 768 et 1536
    est exactement le genre de bug qui produit un classement plausible et faux ;
  * un timeout leve. Il n'est jamais traduit en "ce document n'a pas d'embedding".

Aucune dependance : urllib de la bibliotheque standard suffit pour les deux
fournisseurs, et cela evite d'ajouter `ollama` / `openai` au moteur.
"""
from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Sequence

OLLAMA_MODEL = os.environ.get("TARE_EMBED_MODEL", "granite-embedding:278m")
OLLAMA_DIM = 768
OPENAI_MODEL_DEFAULT = "text-embedding-3-small"
OPENAI_DIMS = {"text-embedding-3-small": 1536, "text-embedding-3-large": 3072,
               "text-embedding-ada-002": 1536}

DEFAULT_BATCH = 16
DEFAULT_TIMEOUT = 120.0


# Duree pendant laquelle Ollama garde le modele en memoire apres un appel.
OLLAMA_KEEP_ALIVE = "30m"

class EmbedError(RuntimeError):
    """L'embedding n'a pas eu lieu. Ce n'est PAS un vecteur nul."""


def _post(url: str, payload: Dict[str, Any], timeout: float,
          headers: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
    body = json.dumps(payload).encode("utf-8")
    h = {"Content-Type": "application/json"}
    h.update(headers or {})
    req = urllib.request.Request(url, data=body, headers=h, method="POST")
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        raw = resp.read()
    try:
        return json.loads(raw)
    except json.JSONDecodeError as exc:  # reponse tronquee : pas un resultat
        raise EmbedError(f"{url}: reponse non-JSON ({len(raw)} octets) — {exc}") from exc


def _get(url: str, timeout: float, headers: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
    req = urllib.request.Request(url, headers=headers or {}, method="GET")
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read())


# ------------------------------------------------------------------ providers

class Embedder:
    provider = "?"
    model = "?"
    dim = 0

    def embed(self, texts: Sequence[str]) -> List[List[float]]:
        raise NotImplementedError

    def embed_one(self, text: str) -> List[float]:
        return self.embed([text])[0]

    def describe(self) -> Dict[str, Any]:
        return {"provider": self.provider, "model": self.model, "dim": self.dim}

    # -- garde commune

    def _check(self, texts: Sequence[str], vectors: Any) -> List[List[float]]:
        if not isinstance(vectors, list) or len(vectors) != len(texts):
            got = len(vectors) if isinstance(vectors, list) else type(vectors).__name__
            raise EmbedError(
                f"{self.provider}/{self.model}: {len(texts)} textes envoyes, {got} vecteurs recus. "
                "Un lot incomplet n'est pas complete par des zeros (regle 3).")
        for i, v in enumerate(vectors):
            if not isinstance(v, list) or len(v) != self.dim:
                raise EmbedError(
                    f"{self.provider}/{self.model}: vecteur {i} de dimension "
                    f"{len(v) if isinstance(v, list) else '?'}, attendu {self.dim}")
            if not any(v):
                raise EmbedError(f"{self.provider}/{self.model}: vecteur {i} entierement nul")
        return [[float(x) for x in v] for v in vectors]


class OllamaEmbedder(Embedder):
    provider = "ollama"

    def __init__(self, url: Optional[str] = None, model: str = OLLAMA_MODEL,
                 dim: int = OLLAMA_DIM, timeout: float = DEFAULT_TIMEOUT,
                 batch: int = DEFAULT_BATCH):
        self.url = (url or os.environ.get("OLLAMA_URL") or "http://127.0.0.1:11434").rstrip("/")
        self.model = model
        self.dim = dim
        self.timeout = timeout
        self.batch = batch

    def available(self) -> tuple:
        try:
            tags = _get(f"{self.url}/api/tags", timeout=5.0)
        except Exception as exc:
            return False, f"{self.url} injoignable: {exc}"
        names = {m.get("name") for m in (tags.get("models") or [])}
        if self.model not in names:
            return False, f"{self.url}: modele {self.model} absent (ollama pull {self.model})"
        return True, ""

    def embed(self, texts: Sequence[str]) -> List[List[float]]:
        if not texts:
            return []
        out: List[List[float]] = []
        for i in range(0, len(texts), self.batch):
            part = list(texts[i:i + self.batch])
            try:
                r = _post(f"{self.url}/api/embed",
                          # keep_alive : sans lui, Ollama decharge le modele apres cinq
                          # minutes d'inactivite et la question suivante paie trente
                          # secondes de rechargement. Le serveur RAG tourne en continu ;
                          # le modele doit y rester.
                          {"model": self.model, "input": part,
                           "keep_alive": OLLAMA_KEEP_ALIVE}, self.timeout)
            except urllib.error.HTTPError as exc:
                raise EmbedError(f"ollama HTTP {exc.code}: {exc.read()[:200]!r}") from exc
            except Exception as exc:
                raise EmbedError(f"ollama {self.url}: {exc}") from exc
            if r.get("error"):
                raise EmbedError(f"ollama: {r['error']}")
            out.extend(self._check(part, r.get("embeddings")))
        return out


class OpenAIEmbedder(Embedder):
    provider = "openai"

    def __init__(self, api_key: Optional[str] = None, model: Optional[str] = None,
                 timeout: float = DEFAULT_TIMEOUT, batch: int = 64,
                 base_url: str = "https://api.openai.com/v1"):
        self.api_key = api_key or os.environ.get("OPENAI_API_KEY") or ""
        self.model = model or os.environ.get("OPENAI_EMBEDDING_MODEL") or OPENAI_MODEL_DEFAULT
        self.dim = OPENAI_DIMS.get(self.model, 1536)
        self.timeout = timeout
        self.batch = batch
        self.base_url = base_url.rstrip("/")

    def available(self) -> tuple:
        return (bool(self.api_key), "" if self.api_key else "OPENAI_API_KEY absente")

    def embed(self, texts: Sequence[str]) -> List[List[float]]:
        if not texts:
            return []
        if not self.api_key:
            raise EmbedError("OPENAI_API_KEY absente")
        out: List[List[float]] = []
        for i in range(0, len(texts), self.batch):
            part = list(texts[i:i + self.batch])
            try:
                r = _post(f"{self.base_url}/embeddings",
                          {"model": self.model, "input": part}, self.timeout,
                          {"Authorization": f"Bearer {self.api_key}"})
            except urllib.error.HTTPError as exc:
                # 429 compris : un rate-limit n'est pas un embedding vide.
                raise EmbedError(f"openai HTTP {exc.code}: {exc.read()[:200]!r}") from exc
            except Exception as exc:
                raise EmbedError(f"openai: {exc}") from exc
            data = r.get("data")
            vecs = [d.get("embedding") for d in data] if isinstance(data, list) else None
            out.extend(self._check(part, vecs))
        return out


# ------------------------------------------------------------------ selection

def make_embedder(prefer: str = "ollama", ollama_url: Optional[str] = None,
                  model: Optional[str] = None) -> tuple:
    """Rend (embedder, journal). Le journal dit POURQUOI ce fournisseur : il
    finit dans le rapport de build, pour qu'on sache six mois plus tard avec
    quel espace vectoriel l'index a ete rempli."""
    log: List[str] = []
    order = ["ollama", "openai"] if prefer != "openai" else ["openai", "ollama"]
    for name in order:
        cand = (OllamaEmbedder(ollama_url, model or OLLAMA_MODEL) if name == "ollama"
                else OpenAIEmbedder(model=model if prefer == "openai" else None))
        ok, why = cand.available()
        if ok:
            log.append(f"{name}: retenu ({cand.model}, {cand.dim} dims)")
            return cand, log
        log.append(f"{name}: ecarte — {why}")
    raise EmbedError("aucun fournisseur d'embeddings disponible. " + " | ".join(log))

File: apps/api/src/rag/embed.ts
/**
 * L'embedding de la REQUETE, cote API.
 *
 * L'index est rempli par engine/tare/rag (granite-embedding:278m via Ollama, 768
 * dimensions). Pour interroger cet index il faut plonger la question dans LE MEME
 * espace vectoriel : un vecteur OpenAI 1536 contre un index Granite 768 ne rend pas
 * un classement moins bon, il ne rend rien du tout. C'est pour cela que
 * `PgVectorStore.search` verifie la dimension AVANT de lancer la requete et refuse
 * plutot que de tronquer ou de completer.
 *
 * Regle 3, ici : un Ollama injoignable, un HTTP 429 ou une reponse tronquee LEVENT.
 * Aucun de ces cas ne se traduit par un vecteur nul, et aucun ne se traduit par
 * `passages: []` — un tableau vide voudrait dire "rien ne correspond", ce qui est
 * une affirmation sur le corpus qu'on n'a pas le droit de faire quand on n'a pas
 * pu poser la question.
 */

export class EmbedError extends Error {
  readonly status: string;
  constructor(message: string, status = "EMBED_FAILED") {
    super(message);
    this.name = "EmbedError";
    this.status = status;
  }
}

export interface EmbedderInfo {
  provider: string;
  model: string;
  dim: number;
}

export interface QueryEmbedder extends EmbedderInfo {
  embed(text: string): Promise<number[]>;
}

const OLLAMA_DIM = 768;
const OPENAI_DIMS: Record<string, number> = {
  "text-embedding-3-small": 1536,
  "text-embedding-3-large": 3072,
  "text-embedding-ada-002": 1536,
};

/** Le vecteur rendu par un fournisseur, verifie avant d'etre utilise. */
function check(v: unknown, dim: number, who: string): number[] {
  if (!Array.isArray(v)) throw new EmbedError(`${who}: pas de vecteur dans la reponse`);
  if (v.length !== dim)
    throw new EmbedError(`${who}: vecteur de dimension ${v.length}, attendu ${dim}`);
  const out = v.map(Number);
  if (out.some((x) => !Number.isFinite(x)))
    throw new EmbedError(`${who}: le vecteur contient une valeur non finie`);
  if (out.every((x) => x === 0)) throw new EmbedError(`${who}: vecteur entierement nul`);
  return out;
}

async function postJson(
  url: string,
  body: unknown,
  timeoutMs: number,
  headers: Record<string, string> = {},
): Promise<unknown> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json", ...headers },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    const text = await res.text();
    if (!res.ok) throw new EmbedError(`${url}: HTTP ${res.status} — ${text.slice(0, 200)}`);
    try {
      return JSON.parse(text);
    } catch {
      // Une reponse tronquee n'est pas une reponse (regle 3, faux resultat #2).
      throw new EmbedError(`${url}: reponse non-JSON (${text.length} octets)`);
    }
  } catch (e) {
    if (e instanceof EmbedError) throw e;
    const msg = (e as Error).name === "AbortError" ? `timeout apres ${timeoutMs} ms` : String(e);
    throw new EmbedError(`${url}: ${msg}`);
  } finally {
    clearTimeout(timer);
  }
}

export class OllamaEmbedder implements QueryEmbedder {
  readonly provider = "ollama";
  constructor(
    readonly baseUrl: string,
    readonly model: string,
    readonly dim: number = OLLAMA_DIM,
    private readonly timeoutMs = 20000,
  ) {}

  async available(): Promise<{ ok: boolean; why: string }> {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 4000);
    try {
      const res = await fetch(`${this.baseUrl}/api/tags`, { signal: ctrl.signal });
      if (!res.ok) return { ok: false, why: `${this.baseUrl}: HTTP ${res.status}` };
      const body = (await res.json()) as { models?: { name?: string }[] };
      const names = new Set((body.models ?? []).map((m) => m.name));
      if (!names.has(this.model))
        return { ok: false, why: `${this.baseUrl}: modele ${this.model} absent` };
      return { ok: true, why: "" };
    } catch (e) {
      return { ok: false, why: `${this.baseUrl} injoignable: ${(e as Error).message}` };
    } finally {
      clearTimeout(timer);
    }
  }

  async embed(text: string): Promise<number[]> {
    const r = (await postJson(
      `${this.baseUrl}/api/embed`,
      { model: this.model, input: text },
      this.timeoutMs,
    )) as { embeddings?: unknown[]; error?: string };
    if (r.error) throw new EmbedError(`ollama: ${r.error}`);
    const first = Array.isArray(r.embeddings) ? r.embeddings[0] : undefined;
    return check(first, this.dim, `ollama/${this.model}`);
  }
}

export class OpenAIEmbedder implements QueryEmbedder {
  readonly provider = "openai";
  readonly dim: number;
  constructor(
    private readonly apiKey: string,
    readonly model: string,
    private readonly timeoutMs = 20000,
    private readonly baseUrl = "https://api.openai.com/v1",
  ) {
    this.dim = OPENAI_DIMS[model] ?? 1536;
  }

  async available(): Promise<{ ok: boolean; why: string }> {
    return this.apiKey
      ? { ok: true, why: "" }
      : { ok: false, why: "OPENAI_API_KEY absente" };
  }

  async embed(text: string): Promise<number[]> {
    if (!this.apiKey) throw new EmbedError("OPENAI_API_KEY absente", "EMBED_UNAVAILABLE");
    const r = (await postJson(
      `${this.baseUrl}/embeddings`,
      { model: this.model, input: text },
      this.timeoutMs,
      { authorization: `Bearer ${this.apiKey}` },
    )) as { data?: { embedding?: unknown }[] };
    const first = Array.isArray(r.data) ? r.data[0]?.embedding : undefined;
    return check(first, this.dim, `openai/${this.model}`);
  }
}

export interface EmbedderChoice {
  embedder: QueryEmbedder;
  log: string[];
}

/**
 * Ollama d'abord (gratuit, local, et c'est le modele qui a rempli l'index),
 * OpenAI en repli. Le journal dit POURQUOI, et il ressort dans /rag/meta : si un
 * jour l'API interroge en 1536 un index rempli en 768, on veut pouvoir le lire.
 */
export async function pickEmbedder(cfg: {
  ollamaUrl: string;
  ollamaModel: string;
  openaiKey: string;
  openaiModel: string;
  prefer?: "ollama" | "openai";
}): Promise<EmbedderChoice> {
  const log: string[] = [];
  const ollama = new OllamaEmbedder(cfg.ollamaUrl.replace(/\/$/, ""), cfg.ollamaModel);
  const openai = new OpenAIEmbedder(cfg.openaiKey, cfg.openaiModel);
  const order = cfg.prefer === "openai" ? [openai, ollama] : [ollama, openai];
  for (const cand of order) {
    const { ok, why } = await cand.available();
    if (ok) {
      log.push(`${cand.provider}: retenu (${cand.model}, ${cand.dim} dims)`);
      return { embedder: cand, log };
    }
    log.push(`${cand.provider}: ecarte — ${why}`);
  }
  throw new EmbedError(
    `aucun fournisseur d'embeddings disponible. ${log.join(" | ")}`,
    "EMBED_UNAVAILABLE",
  );
}
